---
title: "Visualizations"
---

<style type="text/css">
.main-container {
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, error = FALSE, cache = T)
```

```{r}
library(MixSim)
library(RColorBrewer)
library(MASS)
library(expm)
library(grDevices)
library(radviz3d)
library(rgl)
library(ggplot2)
library(Rtsne)
library(Rdimtools)
library(uwot)
library(scatterplot3d) 

source("starcoord3d.R")

# color palette
load("colpal.rda")
dk.palette = c(rgb(red = 173, green = 216, blue = 230, max = 255), 
               rgb(max = 255, red = 139, green = 0, blue = 0))

```

```{r}
##
## Display the map of pairwise overlap measures of Maitra and Melnykov
## (JCGS, 2012)
##
## overlap.mat = matrix of total pairwise overlaps (necessarily symmetric)
## map.col = colormap for the mapping
## linescol = color for the lines drawing the squares
## map.range = range of the overlap map (default: minimum and maximum of lower
##               triangle of the matrix)
## lab.col = color of the labels (same as nrow(matrix) if provided)
## lab.cex = character size of the label
## map.cex = character size of the overlap values laid on the map
## legend.cex = character size of the legend text (does not work always)
##
## provides map of overlap values for each group of mixture model
##
## written Ranjan Maitra, Ames, IA 50011-1210, June 28, 2009
##
## modified Ranjan Maitra, Ames, IA 50011-1090, December 23, 2016.
## last modified Ranjan Maitra, Ames, IA 50011-1090, October 31, 2020.
##
## modification to bring in specifications for label color, labels, maps and
## legend character size; subsequently change default color of map and lines
##

overlap.map <-
    function(overlap.mat, map.col = RColorBrewer::brewer.pal(name = "PuRd", n = 9), linescol = "#1D91C0", map.range = NULL, lab = NULL, lab.col = 1, lab.cex = 5, font = 1, map.cex = 5, legend.cex = 1, scale = NULL, scale.pos = -2, legend.width = 1)
{
    oxmat <- overlap.mat
    oxmat[lower.tri(oxmat)] <- NA
    diag(oxmat) <- NA
    p <- ncol(oxmat)
    if(is.null(lab)){
      lab <- 1:p
    }
    newox <- oxmat[-p, -1]
    newox <- cbind(rbind(NA, newox), NA)[, p:1]

    layout(matrix(c(rep(1, 4*p^2), rep(2, 2*p), rep(3, 2*p), rep(4, 2*p), rep(4, 2*p)), nrow = 2*p, ncol = 2*p + 4 ))
    ##  layout(matrix(c(rep(1, p^2), rep(2, p), rep(3, p), rep(4, p)), nrow = p, ncol = p + 3))

    
    par(mar = c(0.1,0.1,0.75,0.1))
    if (is.null(map.range)) map.range <- range(newox, na.rm = T)  
    image(x = 1:p, y = 1:p, z = newox, axes = F, xlab = "", ylab = "", #col = brewer.pal(9, "GnBu"))
          col = map.col, zlim = map.range)
    text(y = 2:p, x = rep(1, p-1), labels = lab[p:2], cex = lab.cex, font = font, col = lab.col[p:2])
    text(x = 2:p, y = rep(1, (p-1)), labels = lab[1:(p-1)], cex = lab.cex, font = font, col = lab.col[1:(p-1)])
    if (!is.null(scale)){
       text(x = p-scale.pos, y = p+0.2, labels = scale, cex = map.cex)
    }
   


    for(i in 1:p) {
        for(j in i:p) {
            lines(x = c(i+0.5, i+0.5), y = c(p-j+1,p-j)+1.5, col = linescol, lwd = 0.5)
            lines(y = c(i+0.5, i+0.5), x = c(p-j+1,p-j)+1.5, col = linescol, lwd = 0.5)
        }
    }
    for(i in 2:p) text(x=1:p, y = i, labels = round(newox[,i], 2),
                       col = ifelse(newox[,i] < median(map.range), "black", "white"), cex = map.cex)

   frame()

    # savepar <- par(cex=0.75, lwd=0.25, mar = c(1, 0.5, 1, 2),
    #                 xaxs="i", yaxs="i")
    
    savepar <- par(cex=0.75, lwd=0.25, mar = c(0.5, 0.5, 0.5, 0.5),
                   xaxs="i", yaxs="i")
    # if (legend){
    plot.new()
    length.col <- length(map.col) + 1
    ra <- seq(from = map.range[1], to = map.range[2], length=length.col)
    plot.window(xlim=c(0,0.1), ylim= c(map.range[1], map.range[2]))
    rect(0, ra[-length.col], legend.width, ra[-1], col = map.col, border = NULL)
    axis(4, at = ra, labels = round(ra, digits = 2), las = 1, cex.axis = legend.cex, line = NA)
    

    rect(0, 1, legend.width, ra[length.col], col = NULL)
    frame()
    }

```


```{r, eval = F}
# Simulated Datasets
omegas <- c(0.001, 0.01, 0.05)
sim_data <- NULL
overlap_mat_sim <- NULL

seed = runif(3)*10^8
for (i in 1:length(omegas)){
  set.seed(seed[i])
  p = 100
  Q <- MixGOM(goMega = omegas[i], K = 5, p = p, sph = T, hom = T)
  overlap_mat_sim[[i]] <- overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap
  #print(t(overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap) + overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap - diag(5))
  A <- simdataset(n = 500, Pi = Q$Pi, Mu = Q$Mu, S = Q$S)
  sn <- A$X
  class <- as.factor(A$id)
  sn.df <- data.frame(class,sn)
  sim_data[[i]] <- sn.df
}

convert_to_discrete <- function(data, discrete_p = 1:50){
  n <- nrow(data)
  data[,discrete_p] <- apply(data[,discrete_p], MARGIN = 2, FUN = function(x) floor(order(x)/(n+1)*10))
  return(data)
}

for (i in 1:3){
  df <- sim_data[[i]]
  df.mix <- convert_to_discrete(df[,-1])
  sim_data[[i]] <- data.frame(class = df$class, df.mix)
}

save(overlap_mat_sim, sim_data, seed, file = "sim_data_100.rda")

# 2 groups
omegas <- c(0.001, 0.01, 0.05)
sim_data <- NULL
overlap_mat_sim <- NULL

seed = runif(3)*10^8
for (i in 1:length(omegas)){
  set.seed(seed[i])
  p = 100
  Q <- MixGOM(goMega = omegas[i], K = 2, p = p, sph = T, hom = T)
  overlap_mat_sim[[i]] <- overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap
  #print(t(overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap) + overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap - diag(5))
  A <- simdataset(n = 500, Pi = Q$Pi, Mu = Q$Mu, S = Q$S)
  sn <- A$X
  # colnames(sn) <- paste(1:4)
  class <- as.factor(A$id)
  sn.df <- data.frame(class,sn)
  sim_data[[i]] <- sn.df
}

convert_to_discrete <- function(data, discrete_p = 1:50){
  n <- nrow(data)
  data[,discrete_p] <- apply(data[,discrete_p], MARGIN = 2, FUN = function(x) floor(order(x)/(n+1)*10))
  return(data)
}

for (i in 1:3){
  df <- sim_data[[i]]
  df.mix <- convert_to_discrete(df[,-1])
  sim_data[[i]] <- data.frame(class = df$class, df.mix)
}

save(overlap_mat_sim, sim_data, seed, file = "sim_data_100_2g.rda")

# 3 groups
omegas <- c(0.001, 0.01, 0.05)
sim_data <- NULL
overlap_mat_sim <- NULL

seed = runif(3)*10^8
for (i in 1:length(omegas)){
  set.seed(seed[i])
  p = 100
  Q <- MixGOM(goMega = omegas[i], K = 3, p = p, sph = T, hom = T)
  overlap_mat_sim[[i]] <- overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap
  #print(t(overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap) + overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap - diag(5))
  A <- simdataset(n = 500, Pi = Q$Pi, Mu = Q$Mu, S = Q$S)
  sn <- A$X
  # colnames(sn) <- paste(1:4)
  class <- as.factor(A$id)
  sn.df <- data.frame(class,sn)
  sim_data[[i]] <- sn.df
}

convert_to_discrete <- function(data, discrete_p = 1:50){
  n <- nrow(data)
  data[,discrete_p] <- apply(data[,discrete_p], MARGIN = 2, FUN = function(x) floor(order(x)/(n+1)*10))
  return(data)
}

for (i in 1:3){
  df <- sim_data[[i]]
  df.mix <- convert_to_discrete(df[,-1])
  sim_data[[i]] <- data.frame(class = df$class, df.mix)
}

save(overlap_mat_sim, sim_data, seed, file = "sim_data_100_3g.rda")

# 4 groups
omegas <- c(0.001, 0.01, 0.05)
sim_data <- NULL
overlap_mat_sim <- NULL

seed = runif(3)*10^8
for (i in 1:length(omegas)){
  set.seed(seed[i])
  p = 100
  Q <- MixGOM(goMega = omegas[i], K = 4, p = p, sph = T, hom = T)
  overlap_mat_sim[[i]] <- overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap
  #print(t(overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap) + overlap(Q$Pi, Q$Mu, Q$S)$OmegaMap - diag(5))
  A <- simdataset(n = 500, Pi = Q$Pi, Mu = Q$Mu, S = Q$S)
  sn <- A$X
  # colnames(sn) <- paste(1:4)
  class <- as.factor(A$id)
  sn.df <- data.frame(class,sn)
  sim_data[[i]] <- sn.df
}

convert_to_discrete <- function(data, discrete_p = 1:50){
  n <- nrow(data)
  data[,discrete_p] <- apply(data[,discrete_p], MARGIN = 2, FUN = function(x) floor(order(x)/(n+1)*10))
  return(data)
}

for (i in 1:3){
  df <- sim_data[[i]]
  df.mix <- convert_to_discrete(df[,-1])
  sim_data[[i]] <- data.frame(class = df$class, df.mix)
}

save(overlap_mat_sim, sim_data, seed, file = "sim_data_100_4g.rda")
```


## Fig. S1: Overlap Map, starcoords3D and UMAP of Simulated 100D Mixed Datasets with 5 Groups

(a) [Overlap Map](sim_overlap.html)
(b) [starcoords3D](sim_starcoords3d.html)
(c) [UMAP](sim_umap.html)

## Fig. S2: starcoords3D and 2D/3D Scatter Plots of Simulated 100D Mixed Datasets with 2 Groups

(a) [starcoords3D](sim_starcoords3d_2g.html)
(b) [2D Scatter Plot](sim_scatter2d_2g.html)
(c) [3D Scatter Plot](sim_scatter3d_2g.html)

## Fig. S3: Overlap Map, starcoords3D and 2D/3D Scatter Plots of Simulated 100D Mixed Datasets with 3 Groups

(a) [Overlap Map](sim_overlap_3g.html)
(b) [starcoords3D](sim_starcoords3d_3g.html)
(c) [2D Scatter Plot](sim_scatter2d_3g.html)
(d) [3D Scatter Plot](sim_scatter3d_3g.html)

## Fig. S4: Overlap Map, starcoords3D and 2D/3D Scatter Plots of Simulated 100D Mixed Datasets with 4 Groups

(a) [Overlap Map](sim_overlap_4g.html)
(b) [starcoords3D](sim_starcoords3d_4g.html)
(c) [2D Scatter Plot](sim_scatter2d_4g.html)
(d) [3D Scatter Plot](sim_scatter3d_4g.html)


# Real Data Applications

## Fig. S6: Senators Dataset
```{r}
load("datasets-gdt/senators_normal.rda")
senators_normal_class[35,1] = "D"
senators_normal_class$party <- droplevels(senators_normal_class$party)
sn <- as.matrix(senators_normal_class[,-1])
class <- senators_normal_class$party
sn.df <- data.frame(sn,class)

#df.red <- mrp(data = sn.df[,-543], cl = factor(sn.df[,543]), npc = 4)$projected_df
#sndf = data.frame(df.red, class)
#save(sndf, file = "datasets/senatordf.rda")
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
tsne <- Rtsne(as.matrix(sn.df[,-543]), dims = 3, 
              perplexity=min(500,floor((nrow(sn.df)-1)/3)), 
              check_duplicates = F,  
              verbose=F, max_iter = 500)
res1 = tsne$Y
res1 = data.frame(res1, class)
colnames(res1) = c("x","y","z","class")
# open 3d window
 
# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
t-SNE
</td>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.df[,-543]), n_components =  3, y = sn.df[,543]) 

res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
out2 = do.olda(as.matrix(sn.df[,-543]), ndim = 3, class)
senator.olda = data.frame(out2$Y,class)
plot.starcoords2D(senator.olda[,-4], class = senator.olda$class, 
                  colors = dk.palette[as.numeric(senator.olda$class)])
```
<br />
2D Stars Coordinates with OLDA
</td>

<td align="center" valign="center">
```{r}
out2 = do.ulda(as.matrix(sn.df[,-543]), ndim = 3, class)
senator.ulda = data.frame(out2$Y,class)
plot.starcoords2D(senator.ulda[,-4], class = senator.ulda$class, 
                  colors = dk.palette[as.numeric(senator.ulda$class)])
```
<br />
2D Stars Coordinates with ULDA
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-543], cl = factor(sn.df[,543]), npc = 2)$projected_df
ggplot(df.red, aes(x=df.red[,1], y=df.red[,2], pch = factor(sn.df[,543]),
                   color=factor(sn.df[,543]))) + 
    geom_point() + theme_light() + theme(legend.position = "none") + 
    labs(x = "MRP 1", y = "MRP 2") +
    scale_shape_manual(values=20:19) +
    scale_color_manual(values=dk.palette)
```
<br />
2D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-543], cl = factor(sn.df[,543]), npc = 3)$projected_df
scatterplot3d(df.red, pch = (20:1)[as.numeric(class)],
                     color = dk.palette[as.numeric(class)], 
                     xlab = "",  ylab = "", zlab = "", box = F,
                     grid = T, tick.marks = F, label.tick.marks = F)
```
<br />
3D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
load("datasets/senatordf.rda")
plot.starcoords3D(data = sndf[,-5], pradius = 0.02, class = factor(sndf[,5]), 
                  colors = dk.palette)
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>

## Fig. S7: Autisim Dataset
```{r}
load("datasets-gdt/Autism_normal.rda")
sn <- as.matrix(Autism_normal_class[,-1])
class <- Autism_normal_class$class
sn.df <- data.frame(sn,class)
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
tsne <- Rtsne(as.matrix(sn.df[,-15]), dims = 3, 
              perplexity=min(500,floor((nrow(sn.df)-1)/3)), 
              check_duplicates = F,  
              verbose=F, max_iter = 500)
res1 = tsne$Y
res1 = data.frame(res1, class)
colnames(res1) = c("x","y","z","class")
# open 3d window
 
# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
t-SNE
</td>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.df[,-15]), n_components =  3, y = sn.df[,15]) 
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
out2 = do.olda(as.matrix(sn.df[,-15]), ndim = 3, class)
autism.olda = data.frame(out2$Y,class)
plot.starcoords2D(autism.olda[,-4], class = autism.olda$class, 
                  colors = dk.palette[as.numeric(autism.olda$class)])
```
<br />
2D Stars Coordinates with OLDA
</td>

<td align="center" valign="center">
```{r}
out2 = do.ulda(as.matrix(sn.df[,-15]), ndim = 3, class)
autism.ulda = data.frame(out2$Y,class)
plot.starcoords2D(autism.ulda[,-4], class = autism.ulda$class, 
                  colors = dk.palette[as.numeric(autism.ulda$class)])
```
<br />
2D Stars Coordinates with ULDA
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-15], cl = factor(sn.df[,15]), npc = 2)$projected_df
ggplot(df.red, aes(x=df.red[,1], y=df.red[,2], pch = factor(sn.df[,15]),
                   color=factor(sn.df[,15]))) + 
    geom_point() + theme_light() + theme(legend.position = "none") + 
    labs(x = "MRP 1", y = "MRP 2") +
    scale_shape_manual(values=20:19) +
    scale_color_manual(values=dk.palette)
```
<br />
2D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-15], cl = factor(sn.df[,15]), npc = 3)$projected_df
scatterplot3d(df.red, pch = (20:1)[as.numeric(class)],
                     color = dk.palette[as.numeric(class)], 
                     xlab = "",  ylab = "", zlab = "", box = F,
                     grid = T, tick.marks = F, label.tick.marks = F)
```
<br />
3D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
load("datasets/autismdf.rda")
plot.starcoords3D(data = sndf[,-5], pradius = 0.02, class = factor(sndf[,5]), 
                  colors = dk.palette)
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>

## Fig. S8: SPECT Heart Dataset
```{r}
load("datasets-gdt/SPECT_normal.rda")
sn <- as.matrix(SPECT_normal_class[,-1])
class <- as.factor(SPECT_normal_class$class)
sn.df <- data.frame(sn,class)
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
tsne <- Rtsne(as.matrix(sn.df[,-23]), dims = 3, 
              perplexity=min(500,floor((nrow(sn.df)-1)/3)), 
              check_duplicates = F,  
              verbose=F, max_iter = 500)
res1 = tsne$Y
res1 = data.frame(res1, class)
colnames(res1) = c("x","y","z","class")
# open 3d window
 
# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
t-SNE
</td>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.df[,-23]), n_components = 3, y = sn.df[,23])
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = dk.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
out2 = do.olda(as.matrix(sn.df[,-23]), ndim = 3, class)
spect.olda = data.frame(out2$Y,class)
plot.starcoords2D(spect.olda[,-4], class = spect.olda$class, 
                  colors = dk.palette[as.numeric(spect.olda$class)])
```
<br />
2D Stars Coordinates with OLDA
</td>

<td align="center" valign="center">
```{r}
out2 = do.ulda(as.matrix(sn.df[,-23]), ndim = 3, class)
spect.ulda = data.frame(out2$Y,class)
plot.starcoords2D(spect.ulda[,-4], class = spect.ulda$class, 
                  colors = dk.palette[as.numeric(spect.ulda$class)])
```
<br />
2D Stars Coordinates with ULDA
</td>
</tr>

<tr>
<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-23], cl = factor(sn.df[,23]), npc = 2)$projected_df
ggplot(df.red, aes(x=df.red[,1], y=df.red[,2], pch = factor(sn.df[,23]),
                   color=factor(sn.df[,23]))) + 
    geom_point() + theme_light() + theme(legend.position = "none") + 
    labs(x = "MRP 1", y = "MRP 2") +
    scale_shape_manual(values=20:19) +
    scale_color_manual(values=dk.palette)
```
<br />
2D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.red <- mrp(data = sn.df[,-23], cl = factor(sn.df[,23]), npc = 3)$projected_df
scatterplot3d(df.red, pch = (20:1)[as.numeric(class)],
                     color = dk.palette[as.numeric(class)], 
                     xlab = "",  ylab = "", zlab = "", box = F,
                     grid = T, tick.marks = F, label.tick.marks = F)
```
<br />
3D Scatter Plot
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
load("datasets/spectdf.rda")
plot.starcoords3D(data = sndf[,-5], pradius = 0.02, class = factor(sndf[,5]), 
                  colors = dk.palette)
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>


## Fig. S9: Indic Scripts Dataset
```{r}
load("datasets-gdt/indic_normal.rda")
sn <- as.matrix(indic_normal_class[,-1])
class <- as.factor(indic_normal_class$class)
script <- data.frame(sn,class)
sn.df = script
g = c(1,3,4,5,6,11)
sn.g.df = sn.df[sn.df$class %in% g, ]
class = droplevels(sn.g.df$class)

script.palette <- brewer.pal(12,"Paired")[-c(2,7)]
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
tsne <- Rtsne(as.matrix(sn.g.df[,-117]), dims = 3, 
              perplexity=min(500,floor((nrow(sn.g.df)-1)/3)), 
              check_duplicates = F,  
              verbose=F, max_iter = 500)
res1 = tsne$Y
res1 = data.frame(res1, class)
colnames(res1) = c("x","y","z","class")
# open 3d window
 
# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = script.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
t-SNE
</td>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.g.df[,-117]), n_components = 3, y = sn.g.df[,117])
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = script.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
out2 = do.olda(as.matrix(sn.g.df[,-117]), ndim = 4, class)
script.olda = data.frame(out2$Y,class)
plot.starcoords2D(script.olda[,-5], class = script.olda$class, 
                  colors = dk.palette[as.numeric(script.olda$class)])
```
<br />
2D Stars Coordinates with OLDA
</td>

<td align="center" valign="center">
```{r}
out2 = do.ulda(as.matrix(sn.g.df[,-117]), ndim = 4, class)
script.ulda = data.frame(out2$Y,class)
plot.starcoords2D(script.ulda[,-5], class = script.ulda$class, 
                  colors = dk.palette[as.numeric(script.ulda$class)])
```
<br />
2D Stars Coordinates with ULDA
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
load("datasets/scriptdf.rda")
ram.ord = c(2,3,4,1)
plot.starcoords3D(data = scriptdf[,ram.ord], pradius = 0.06, 
                  class = factor(scriptdf[,5]), colors = script.palette)
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>

## Fig. S10: Gtex Dataset
```{r}
load("datasets-gdt/GTEX_normal.rda")
sn <- as.matrix(d.red_normal)
class <- as.factor(as.numeric(as.factor(group)))
class = as.factor(gsub("13","3",class))
sn.df = data.frame(sn,class)
g = c(2,4,5,7,8,9,11,12)
sn.g.df = sn.df[sn.df$class %in% g, ]
class = droplevels(sn.g.df$class)
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.g.df[,-20243]), n_components = 3, 
                 y = sn.g.df[,20243], target_weight = 0.5)
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = script.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
load("datasets/gtexdf.rda")
plot.starcoords3D(data = sndf[,-8], pradius = 0.02, class = factor(sndf[,8]), 
                  colors = gtex.palette)
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>

## Fig. S11: HCV Sequence Dataset
```{r, eval = F}
# data preprocessing
library(stringi)
xx <- ampir::read_faa(file = "HCV_ALL_2008_genome_DNA.fasta")
# create classes
# The first digit: genotype (7 of them); the second letter: subtype. 
genotype = genosub <- rep(NA, nrow(xx))
for (i in 1:nrow(xx)) {
  genotype[i] = substr(xx[i,1],1,1)
  genosub[i] = substr(xx[i,1],1,2)
}
genotype <- factor(genotype)
genosub <- factor(genosub)

seq.p = NULL # dimensions for each obs.
for (i in 1:nrow(xx)) {
  seq.p = c(seq.p, nchar(xx[i,2]))
}

# 2-bits coding: A, G, C, T -> 00, 01, 10, 11; R,S,Y,-, etc -> missing
data <- NULL
for (i in 1:nrow(xx)) {
  str.trim <- substr(xx[i,2],1,min(seq.p))
  str.trim <- stri_replace_all_regex(str.trim,
                                  pattern=c('A', 'C', 'G', 'T'),
                                  replacement=c('00', '01', '10', '11'),
                                  vectorize=FALSE)
  str.trim <- gsub("[^0-9.-]", "22", str.trim)
  str.trim <- gsub("-", "22", str.trim)
  vec <- unlist(strsplit(str.trim,""))
  vec[vec=="2"] <- NA
  vec <- as.numeric(vec)
  data <- rbind(data, vec)
}
data <- data[, colSums(is.na(data)) == 0]
dim(data) #471 7943*2
hcv.2bits <- data.frame(genotype, genosub, data)

# 4 bits coding: A, G, C, T -> 0001, 0010, 0100, 1000
data <- NULL
for (i in 1:nrow(xx)) {
  str.trim <- substr(xx[i,2],1,min(seq.p))
  str.trim <- stri_replace_all_regex(str.trim,
                                  pattern=c('A', 'C', 'G', 'T'),
                                  replacement=c('0001', '0010', '0100', '1000'),
                                  vectorize=FALSE)
  str.trim <- gsub("[^0-9.-]", "2222", str.trim)
  str.trim <- gsub("-", "2222", str.trim)
  vec <- unlist(strsplit(str.trim,""))
  vec[vec=="2"] <- NA
  vec <- as.numeric(vec)
  data <- rbind(data, vec)
}
data <- data[, colSums(is.na(data)) == 0]
dim(data) #471 7943*4
hcv.4bits <- data.frame(genotype, genosub, data)

save(hcv.2bits,hcv.4bits, file = "datasets/hcvseq.rda")
```

```{r}
load("datasets/hcvseq.rda")
summary(hcv.2bits$genotype)
# remove groups 5 and 7, and the genosub column
index <- c(which(hcv.2bits$genotype=="5"), which(hcv.2bits$genotype=="7"))

df = hcv.4bits[-index,-2]
df$genotype = droplevels(df$genotype)
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.g.df[,-20243]), n_components = 3, 
                 y = sn.g.df[,20243], target_weight = 0.5)
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = script.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.gdt <- radviz3d::Gtrans(df[,-1])
# for 4 directions, minimum of 3 observation(s) in a cluster.
df.red <- mrp(data = df.gdt, cl = factor(df$genotype), npc = 4)$projected_df
plot.starcoords3D(data = df.red, class = factor(df$genotype), cex = 1)
rgl::rgl.viewpoint(zoom = 0.6)
rgl::rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>

## Fig. S12: HIV Dataset
```{r, eval = F}
# data preprocessing
library(stringi)
xx <- ampir::read_faa(file = "for_yf.fsa")
# create classes
# The first digit: genotype (7 of them); the second letter: subtype. 
class <- rep(NA, nrow(xx))
for (i in 1:nrow(xx)) {
  class[i] = substr(xx[i,1],1,2)
}
class <- factor(class)

seq.p = NULL # dimensions for each obs.
for (i in 1:nrow(xx)) {
  seq.p = c(seq.p, nchar(xx[i,2]))
}

# 2-bits coding: A, G, C, T -> 00, 01, 10, 11; R,S,Y,-, etc -> missing
data <- NULL
for (i in 1:nrow(xx)) {
  str.trim <- substr(xx[i,2],1,min(seq.p))
  str.trim <- stri_replace_all_regex(str.trim,
                                  pattern=c('A', 'C', 'G', 'T'),
                                  replacement=c('00', '01', '10', '11'),
                                  vectorize=FALSE)
  str.trim <- gsub("[^0-9.-]", "22", str.trim)
  str.trim <- gsub("-", "22", str.trim)
  vec <- unlist(strsplit(str.trim,""))
  vec[vec=="2"] <- NA
  vec <- as.numeric(vec)
  data <- rbind(data, vec)
}
data <- data[, colSums(is.na(data)) == 0]
dim(data) #73 7432*2
hiv.2bits <- data.frame(class, data)

# 4 bits coding: A, G, C, T -> 0001, 0010, 0100, 1000
data <- NULL
for (i in 1:nrow(xx)) {
  str.trim <- substr(xx[i,2],1,min(seq.p))
  str.trim <- stri_replace_all_regex(str.trim,
                                  pattern=c('A', 'C', 'G', 'T'),
                                  replacement=c('0001', '0010', '0100', '1000'),
                                  vectorize=FALSE)
  str.trim <- gsub("[^0-9.-]", "2222", str.trim)
  str.trim <- gsub("-", "2222", str.trim)
  vec <- unlist(strsplit(str.trim,""))
  vec[vec=="2"] <- NA
  vec <- as.numeric(vec)
  data <- rbind(data, vec)
}
data <- data[, colSums(is.na(data)) == 0]
dim(data) #73 7432*4
hiv.4bits <- data.frame(class, data)

save(hiv.2bits,hiv.4bits, file = "datasets/hiv.rda")
```

```{r}
load("datasets/hcvseq.rda")
summary(hcv.2bits$genotype)
# remove groups 5 and 7, and the genosub column
index <- c(which(hcv.2bits$genotype=="5"), which(hcv.2bits$genotype=="7"))

df = hcv.4bits[-index,-2]
df$genotype = droplevels(df$genotype)
```


<table width="500" border="0" cellpadding="0">

<tr>

<td align="center" valign="center">
```{r}
set.seed(1)
umap_res <- umap(as.matrix(sn.g.df[,-20243]), n_components = 3, 
                 y = sn.g.df[,20243], target_weight = 0.5)
res1 = data.frame(umap_res, class)
colnames(res1) = c("x","y","z","class")
# open 3d window

# resize window
par3d(windowRect = c(100, 100, 612, 612))
# plot
with(res1, plot3d(xlab = "", ylab="", zlab="",axes = F, x,y,z, type="s",size = 0.75,
                  col = script.palette[1:nlevels(class)][as.numeric(class)]))
box3d()
rgl::rgl.viewpoint(zoom = 0.6)
rglwidget()
```
<br />
UMAP
</td>
</tr>

<tr>

<td align="center" valign="center">
```{r}
df.gdt <- radviz3d::Gtrans(df[,-1])
# for 4 directions, minimum of 3 observation(s) in a cluster.
df.red <- mrp(data = df.gdt, cl = factor(df$genotype), npc = 4)$projected_df
plot.starcoords3D(data = df.red, class = factor(df$genotype), cex = 1)
rgl::rgl.viewpoint(zoom = 0.6)
rgl::rglwidget()
```
<br />
Starcoords3D
</td>
</tr>
</table>







